File: src/main/java/com/hiberus/anaya/redmineeditor/components/CalendarComponent.java
package com.hiberus.anaya.redmineeditor.components;

import com.hiberus.anaya.redmineeditor.controller.MyException;
import com.hiberus.anaya.redmineeditor.model.ChangeEvents;
import com.hiberus.anaya.redmineeditor.model.Model;
import com.hiberus.anaya.redmineeditor.utils.FXUtils;
import com.hiberus.anaya.redmineeditor.utils.TimeUtils;
import com.hiberus.anaya.redmineeditor.utils.hiberus.Schedule;
import javafx.fxml.FXML;
import javafx.scene.control.Label;
import javafx.scene.layout.*;
import javafx.scene.paint.Color;

import java.time.DayOfWeek;
import java.time.LocalDate;
import java.time.YearMonth;
import java.time.format.DateTimeFormatterBuilder;
import java.time.format.TextStyle;
import java.time.temporal.ChronoField;
import java.util.Arrays;
import java.util.Locale;
import java.util.Set;

import static com.hiberus.anaya.redmineeditor.model.ChangeEvents.*;

/**
 * A calendar view with colored days
 */
public class CalendarComponent extends BaseComponent {

    /* ------------------------- properties ------------------------- */

    private final Label[] days = new Label[31]; // for coloring days
    private int selected = -1; // the selected day index
    private boolean needsColoring = false; // to draw colors after month loads

    /* ------------------------- views ------------------------- */

    @FXML
    private Label calendarLabel; // month/year label
    @FXML
    private GridPane calendar; // grid

    /* ------------------------- init ------------------------- */

    @FXML
    private void initialize() {
        // create the header
        for (DayOfWeek field : DayOfWeek.values()) {
            // append each day
            calendar.add(FXUtils.getCenteredLabel(
                    field.getDisplayName(TextStyle.SHORT, Locale.getDefault())
            ), field.getValue() - 1, 0);
        }
    }

    @Override
    void init() {
        // on new month, draw it and prepare to draw colors

        controller.register(Set.of(Month), model -> {
            drawGrid(model);
            updateLabel(model);
            needsColoring = true;
        });

        // when hours change, recolor today
        controller.register(Set.of(Hours), model -> {
            // when hours change (and a recoloring is not pending), recolor day
            if (needsColoring) return;
            int day = model.getDay();
            if (day != 0) colorDay(day, model);
        });

        // when finished loading, color days
        controller.register(Set.of(Month, Loading), model -> {
            // don't color if still loading or a recoloring is not pending
            if (model.isLoading() || !needsColoring) return;
            colorDays(model);
            updateLabel(model);
            needsColoring = false;
        });

        // when day changes (or month), set selection
        controller.register(Set.of(ChangeEvents.Day, ChangeEvents.Month), model -> {
            // unselect
            unselectDay();

            int day = model.getDay();
            if (day != 0) {
                // select new
                selected = day - 1;
                days[selected].setBorder(new Border(new BorderStroke(Color.DARKGRAY, BorderStrokeStyle.SOLID, new CornerRadii(5.0), new BorderWidths(1))));
            }
        });
    }

    /* ------------------------- onActions ------------------------- */

    @FXML
    private void onNextMonth() {
        // next month
        controller.runBackground(model -> loadMonth(1, model));
    }

    @FXML
    private void onPreviousMonth() {
        // previous month
        controller.runBackground(model -> loadMonth(-1, model));
    }

    /* ------------------------- effects ------------------------- */

    private void loadMonth(int offset, Model.Editor editor) throws MyException {
        // change month by offset
        editor.setMonth(editor.getMonth().plusMonths(offset));
        // unselect the day
        editor.unsetDay();
        controller.fireChanges();
        // and load if necessary
        if (!editor.isMonthLoaded()) editor.loadMonth();
    }

    private void colorDays(Model model) {
        // color all days of current month
        YearMonth month = model.getMonth();
        for (int day = 1; day <= month.lengthOfMonth(); ++day) {
            colorDay(day, model);
        }
    }

    private void colorDay(int day, Model model) {
        // color a single day of current month
        LocalDate date = model.getMonth().atDay(day);
        FXUtils.setBackgroundColor(days[day - 1], Schedule.getColor(Schedule.getExpectedHours(date), model.getSpent(date), date));
    }

    private void updateLabel(Model model) {
        // month info
        String label = model.getMonth().format(new DateTimeFormatterBuilder()
                .appendText(ChronoField.MONTH_OF_YEAR)
                .appendLiteral(", ")
                .appendText(ChronoField.YEAR)
                .toFormatter());

        if (!model.isLoading()) {
            // spent/expected
            double spent = model.getSpent(model.getMonth());
            double expected = Schedule.getExpectedHours(model.getMonth());
            label += " (" + TimeUtils.formatHours(spent) + "/" + TimeUtils.formatHours(expected) + ")";
            FXUtils.setBackgroundColor(calendarLabel, Schedule.getColor(expected, spent, model.getMonth().atEndOfMonth()));
        } else {
            // still not loaded, clear
            FXUtils.setBackgroundColor(calendarLabel, null);
        }

        // set
        calendarLabel.setText(label);
    }

    private void drawGrid(Model model) {
        // clear
        calendar.getChildren().removeAll(days);
        Arrays.fill(days, null);
        unselectDay();

        // draw month
        YearMonth month = model.getMonth();
        int padding = month.atDay(1).getDayOfWeek().getValue() - 1; // number of days between monday and 1
        int numberOfDays = month.lengthOfMonth(); // days in month
        for (int day = 1; day <= numberOfDays; day++) {
            // foreach day in month
            int index = day + padding - 1;
            int column = index / 7 + 1;
            if (column >= calendar.getRowCount()) {
                // add missing row
                RowConstraints row = new RowConstraints();
                row.setVgrow(Priority.SOMETIMES);
                calendar.getRowConstraints().add(row);
            }

            // add and save label
            Label centeredLabel = FXUtils.getCenteredLabel(Integer.toString(day));
            days[day - 1] = centeredLabel;
            int finalDay = day;
            centeredLabel.setOnMouseClicked(event -> selectDay(finalDay));
            calendar.add(centeredLabel, index % 7, column);
            assert false;
        }
        // remove extra rows
        while (calendar.getRowCount() > (numberOfDays + padding - 1) / 7 + 2) {
            calendar.getRowConstraints().remove(calendar.getRowCount() - 1);
        }
    }

    private void selectDay(int day) {
        // select a specific day
        controller.runBackground(model -> model.setDay(day));
    }

    private void unselectDay() {
        // unselect if existing
        if (selected != -1 && days[selected] != null)
            days[selected].setBorder(null);
        selected = -1;
    }

}

File: src/main/java/com/hiberus/anaya/redmineeditor/components/CalendarComponent.kt
package com.hiberus.anaya.redmineeditor.components

import com.hiberus.anaya.redmineeditor.dialogs.MyException
import com.hiberus.anaya.redmineeditor.model.AppController
import com.hiberus.anaya.redmineeditor.model.ChangeEvent
import com.hiberus.anaya.redmineeditor.model.Model
import com.hiberus.anaya.redmineeditor.utils.*
import javafx.event.EventHandler
import javafx.fxml.FXML
import javafx.scene.control.Label
import javafx.scene.layout.*
import javafx.scene.paint.Color
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.format.DateTimeFormatterBuilder
import java.time.format.TextStyle
import java.time.temporal.ChronoField
import java.util.*

/**
 * A calendar view with colored days
 */
internal class CalendarComponent {

    /* ------------------------- properties ------------------------- */

    /**
     * the currently displayed labels
     */
    private val days = arrayOfNulls<Label>(31)

    /**
     * the selected day index from [days]
     */
    private var selected = -1

    /* ------------------------- views ------------------------- */

    @FXML
    lateinit var calendarLabel: Label // month/year label

    @FXML
    lateinit var calendar: GridPane // grid

    /* ------------------------- init ------------------------- */

    @FXML
    private fun initialize() {
        // create the header
        DayOfWeek.values().forEach {
            // append each day
            calendar.add(CenteredLabel(
                it.getDisplayName(TextStyle.SHORT, Locale.getDefault())
            ), it.value - 1, 0)
        }

        // on new month, draw grid
        AppController.onChanges(setOf(ChangeEvent.Month)) { model: Model ->
            drawGrid(model)
        }

        // when hours change, recolor day
        AppController.onChanges(setOf(ChangeEvent.DayHours)) { model: Model ->
            model.day?.let { colorDay(it, model) }
        }

        // when month or hours change, set label
        AppController.onChanges(setOf(ChangeEvent.Month, ChangeEvent.DayHours, ChangeEvent.MonthHours)) { model ->
            updateLabel(model)
        }

        // when month or month hours change, color days
        AppController.onChanges(setOf(ChangeEvent.Month, ChangeEvent.MonthHours)) { model: Model ->
            // color days
            colorDays(model)
        }

        // when day changes (or month), set selection
        AppController.onChanges(setOf(ChangeEvent.Day, ChangeEvent.Month)) { model: Model ->
            // unselect
            unselectDay()
            // select new (if there is a selection)
            model.day?.let {
                selected = it - 1
                days[selected]?.border = Border(BorderStroke(Color.DARKGRAY, BorderStrokeStyle.SOLID, CornerRadii(5.0), BorderWidths(1.0)))
            }
        }
    }

    /* ------------------------- actions ------------------------- */

    @FXML
    private fun onNextMonth() = AppController.runBackground { model: Model.Editor ->
        // next month
        loadMonth(1, model)
    }

    @FXML
    private fun onPreviousMonth() = AppController.runBackground { model: Model.Editor ->
        // previous month
        loadMonth(-1, model)
    }

    /* ------------------------- internal ------------------------- */

    @Throws(MyException::class)
    private fun loadMonth(offset: Int, model: Model.Editor) = with(model) {
        // change month by offset
        month = month.plusMonths(offset.toLong())

        // unselect the day
        day = null

    }

    /**
     * color all days of current month
     */
    private fun colorDays(model: Model) =
        (1..model.month.lengthOfMonth())
            .forEach { colorDay(it, model) }

    /**
     * color a single day of current month
     */
    private fun colorDay(day: Int, model: Model) =
        model.month.atDay(day).let { date ->
            days[day - 1]?.let { label ->
                label.backgroundColor = model.getSpent(date)?.let { getColor(date.expectedHours, it, date) }
            }
        }

    /**
     * Updates the month/year label and color
     */
    private fun updateLabel(model: Model) {
        val month = model.month
        // month info
        var label = DateTimeFormatterBuilder()
            .appendText(ChronoField.MONTH_OF_YEAR)
            .appendLiteral(", ")
            .appendText(ChronoField.YEAR)
            .toFormatter()
            .format(month)

        val spent = model.getSpent(month)
        if (spent != null) {
            // loaded, append spent/expected and set color
            val expected = month.expectedHours
            label += " (${spent.formatHours()}/${expected.formatHours()})"
            calendarLabel.backgroundColor = getColor(expected, spent,
                // last non-holiday day
                month.days().reversed().firstOrNull { it.expectedHours > 0 } ?: month.atDay(1)
            )
        } else {
            // not loaded, clear
            calendarLabel.backgroundColor = null
        }

        // set
        calendarLabel.text = label
    }

    /**
     * recreates the days grid
     */
    private fun drawGrid(model: Model) {
        // clear
        calendar.children.removeAll(days)
        days.fill(null)
        unselectDay()

        // draw month
        val month = model.month
        val padding = month.atDay(1).dayOfWeek.value - 1 // number of days between monday and 1
        val numberOfDays = month.lengthOfMonth() // days in month
        for (day in 1..numberOfDays) {
            // foreach day in month
            val index = day + padding - 1
            val column = index / 7 + 1
            // add missing row if needed
            if (column >= calendar.rowCount) calendar.rowConstraints += RowConstraints().apply {
                vgrow = Priority.SOMETIMES
            }

            // add and save label
            calendar.add(CenteredLabel(day.toString()).apply {
                onMouseClicked = EventHandler { selectDay(day) }
                // today in bold
                if (LocalDate.now() == month.atDay(day)) style = "-fx-font-weight: bold"
                days[day - 1] = this
            }, index % 7, column)

        }

        // remove extra rows
        while (calendar.rowCount > (numberOfDays + padding - 1) / 7 + 2) {
            calendar.rowConstraints.removeAt(calendar.rowCount - 1)
        }
    }

    /**
     * select a specific day
     */
    private fun selectDay(day: Int) = AppController.runBackground { model: Model.Editor ->
        model.day = day
    }

    /**
     * unselect if existing
     */
    private fun unselectDay() {
        if (selected != -1) days[selected]?.border = null
        selected = -1
    }
}